File: leetcode/202_happy_number.py
from functools import lru_cache


class Solution:

    HAPPY = 1

    def isHappy(self, n: int) -> bool:
        slow = fast = n
        while True:
            slow = self.get_next(slow)
            fast = self.get_next(self.get_next(fast))
            if slow == self.HAPPY or fast == self.HAPPY:
                return True
            if slow == fast:
                return False

    @lru_cache(None)
    def get_next(self, n: int) -> int:
        next_num = 0
        while n:
            n, last_digit = divmod(n, 10)
            next_num += last_digit ** 2
        return next_num

File: leetcode/java/202HappyNumber.java
/**
 * T: O(logN) S: O(logN)
 *
 * <p>Perform the computation described in the program, remember seen numbers. If we detect a loop -
 * the number is not happy. Otherwise, the number will end up at 1.
 */
import java.util.HashSet;
import java.util.Set;

class Solution {
  public boolean isHappy(int n) {
    Set<Integer> seenNumbers = new HashSet<>();
    while (n != 1) {
      if (seenNumbers.contains(n)) {
        return false;
      }
      seenNumbers.add(n);
      int nextNum = 0;
      while (n != 0) {
        int currentDigit = n % 10;
        nextNum += currentDigit * currentDigit;
        n /= 10;
      }
      n = nextNum;
    }
    return true;
  }
}
